[EEG] Add view/edit functionality and permissions for EEG BIDS derivatives

AlexandraLivadas opened the pull request:
1. Added a new permission for a user to view and/or edit annotations for specific entries in the electrophysiology browser module. 
2. Created functions in the backend of the electrophysiology browser module to get and display derivative data from the database and to write derivative data to the database. There is no view/edit/delete functionality in the front-end added by this PR.
3. Added the ability to download annotation files for a given session. Created a function that updates the annotation files to match changes in the database before they are downloaded by a user. This function can be called by a cronjob using the `tools/update_annotation_files.php` to update all the files periodically.

**Note:** The `tools/updateDerivativesTest` was added for testing purposes and should be removed before merging.  

#### Testing instructions

To test the SQL patch: 
1. Source the `2021-04-23-annotation_permissions.sql` patch or refresh the database with the edited SQL permissions schema and see that the correct entries are added to the database.

To test the download functionality:
1. Check that your database includes some derivative datasets and files. 
2. Navigate to the 
3. See that the button to download annotation files is visible and clickable
4. Check that clicking this button correctly downloads a .tgz file with the correct annotation files
5. Add new data to the `physiological_annotation_instance` and `physiological_annotation_parameter` tables for the respective files. 
6. Re-download the files and see that they are correctly updated with the new information.

Related to #7345

laemtl commented on modules/electrophysiology_browser/php/sessions.class.inc:
```diff
@@ -53,6 +53,32 @@ class Sessions extends \NDB_Page
+     */
+    function _hasViewDerivPerms(\User $user) : bool
+    {
+        return ($user->hasPermission('electrophysiology_browser_view_annotations'));
```
I think you're missing the superuser case :)

AlexandraLivadas replied:
I think the superuser case is handled in the `hasPermission` method at [this line](https://github.com/aces/Loris/blob/2fceb83b8f11bc2b315e279d260730e3d11a40a5/php/libraries/UserPermissions.class.inc#L110)

laemtl replied:
True! My mistake, it's all good then.

laemtl commented on modules/electrophysiology_browser/php/sessions.class.inc:
```diff
@@ -53,6 +53,32 @@ class Sessions extends \NDB_Page
+     */
+    function _hasEditDerivPerms(\User $user) : bool
+    {
+        return ($user->hasPermission('electrophysiology_browser_edit_annotations'));
```
Same here!

christinerogers commented:
Hi Alex, To align with the Imaging Browser permissions, I'd suggest just adding one permission to Add annotations. 
The capacity to Edit/Delete annotations can be tackled after the user can first Add annotations (top priority) 

cc @cmadjar any preferences here?

cmadjar commented:
@christinerogers I may be wrong but I think on the imaging side, when a person can do QC, they can edit later on the QC values as well. Any reason why there should be a separation between add/edit/delete for EEG?

christinerogers commented:
> @christinerogers I may be wrong but I think on the imaging side, when a person can do QC, they can edit later on the QC values as well. Any reason why there should be a separation between add/edit/delete for EEG?

Hi Cecile - agreed, they should be governed by the same permissions, I was just thinking that the functionality to edit an existing annotation will be added after the Add Annotation functionality (one step at a time).

AlexandraLivadas commented:
@christinerogers I removed the `View` permissions (as those are covered by the pre-existing permissions for the module) but kept the `Create/Edit`permission as is. For now, I will only be writing the functionality to add new annotations and not to edit existing ones.

christinerogers commented on modules/electrophysiology_browser/README.md:
```diff
@@ -16,20 +17,24 @@ been correctly inserted into LORIS.
 electrophysiology_browser_view_site
   - This permission gives the user access to electrophysiology datasets from their own site(s) only.
+  
+electrophysiology_browser_edit_annotations
```
perfect. 👍

christinerogers commented on modules/electrophysiology_browser/README.md:
```diff
@@ -16,20 +17,24 @@ been correctly inserted into LORIS.
-sufficient to provide access to view data in the module.
+The Electrophysiology Browser uses the following permissions. Either of the first two is 
+sufficient to provide access to view data in the module. The third permission provides editing
+permissions for derivative datasets.
```
```suggestion
permissions to add or modify annotations (derivatives) for data from all sites in this module.
```
@AlexandraLivadas  can you update these words `from all sites` if not accurate ?  does your permission to add/edit on all-sites or just the user's affiliated sites?

AlexandraLivadas replied:
I will update this because it does not apply for all-sites. The 2 permissions that existed before determine if the session can be viewed and this permission just determines if the annotations of the session being viewed can be edited.

christinerogers commented on modules/electrophysiology_browser/README.md:
```diff
@@ -16,20 +17,24 @@ been correctly inserted into LORIS.
 The Electrophysiology Browser displays electrophysiology datasets that have been
 inserted into LORIS from a BIDS-format collection. Derived or processed electrophysiology
-datasets can also be accessed via this module.
+datasets can also be accessed and edited via this module.
```
```suggestion
datasets can also be accessed and annotated via this module.
```
annotated not edited.

christinerogers commented on modules/electrophysiology_browser/README.md:
```diff
@@ -3,7 +3,8 @@
 The Electrophysiology Browser is intended to allow users to view candidate
-electrophysiology (EEG, MEG...) sessions collected for a study.
+electrophysiology (EEG, MEG...) sessions collected for a study and any associated
+annotations or derivative datasets for the session.
```
```suggestion
annotations (derivatives) for each recording.
```
updating for accuracy

christinerogers commented on modules/electrophysiology_browser/README.md:
```diff
@@ -16,20 +17,24 @@ been correctly inserted into LORIS.
 
-You can download all the files related to a recording (channel information,
-electrode information, task event information, the actual recording...).
+You can download all the files related to a recording and its derivatives (channel information,
+electrode information, task event information, the actual recording, the annotations and their metadata...).
```
```suggestion
You can download all the files related to a recording (channel information,
electrode information, task event information, the actual recording) -- as well as its derivatives (annotations and their metadata).
```
updated this for clarity.  also no ellipses in documentation if you can help it...

christinerogers reviewed (changes requested):
Alex - nice documentation update. 
Can you accept my changes and I left you one fact-check to followup on, inline below. 

More rainy day documetnation updates for this module will include: 
- [ ] Help text 
- [ ] updating when edit/delete actually become feasible.

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -18,7 +19,9 @@ class FilePanel extends Component {
     super(props);
     this.state = {
       data: this.props.data,
+      id: this.props.fileId,
```
For clarity, can this key and the props be renamed for something more explicit (like physiologicalFileID?)

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -248,6 +251,7 @@ class FilePanel extends Component {
                   >
                     <button id='btn_download_all_files'
                             style={stylesFile.button.download}
+                            onClick={this.handleClick}
```
handleClick is too vague maybe? I think something more descriptive could help (like annotationsUpdate).

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -555,16 +560,85 @@ class FilePanel extends Component {
+     * update the annotation files before downloading
+     */
+    handleClick() {
+        console.log('The link was clicked.');
```
Oups :) console.log can be removed.

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -555,16 +560,85 @@ class FilePanel extends Component {
+        console.log(this.state.id);
+
+        fetch(this.props.url, {
+            method: 'PUT',
```
I believe PUT is semantically incorrect: https://www.guru99.com/put-vs-post.html#5
This operation is not idempotent.

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -555,16 +560,85 @@ class FilePanel extends Component {
+                let msgType = 'success';
+                let message = 'Your files will be downloaded in 2 seconds!';
+                this.showAlertMessage(msgType, message);
+                console.log(message);
```
Not needed?

laemtl replied:
Do you have cases where an error can occur? For example, if the physiologicalID is not found.

AlexandraLivadas replied:
Yes there are cases where the file is not found or the file path is incorrect/not found as well

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -555,16 +560,85 @@ class FilePanel extends Component {
+     */
+    showAlertMessage(msgType, message) {
+        let type = 'success';
+        let title = 'Issue updated!';
```
Issue?

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -555,16 +560,85 @@ class FilePanel extends Component {
+    showAlertMessage(msgType, message) {
+        let type = 'success';
+        let title = 'Issue updated!';
+        let text = message || '';
```
text can be a const.

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -555,16 +560,85 @@ class FilePanel extends Component {
+     * @param {string} message - message content
+     */
+    showAlertMessage(msgType, message) {
+        let type = 'success';
```
For variables with different values, I think it's better to define here and assign either in the if or else case.

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -555,16 +560,85 @@ class FilePanel extends Component {
+            type: type,
+            text: text,
+            timer: timer,
+            allowOutsideClick: false,
```
allowOutsideClick and allowEscapeKey are convenient if the user does not need to make a choice. Why setting them to false?

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -555,16 +560,85 @@ class FilePanel extends Component {
   id: 'file_panel',
   title: 'FILENAME',
   data: {},
+  url: '',
```
url should not have a default value I think.

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -555,16 +560,85 @@ class FilePanel extends Component {
+    }
 }
+
 FilePanel.propTypes = {
```
Is fileId missing?

christinerogers replied:
@AlexandraLivadas Is this comment outstanding or addressed / resolved?

laemtl commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -555,16 +560,85 @@ class FilePanel extends Component {
   id: PropTypes.string,
   title: PropTypes.string,
   data: PropTypes.object,
+  url: PropTypes.string,
```
I think the standard name for this is action.

laemtl commented on modules/electrophysiology_browser/jsx/electrophysiologySessionView.js:
```diff
@@ -262,6 +263,10 @@ class ElectrophysiologySessionView extends Component {
               id={'filename_panel_' + i}
               title={this.state.database[i].file.name}
               data={this.state.database[i].file}
+              fileId={this.state.database[i].file.id}
```
Should we rename fileID for something more descriptive?

christinerogers commented on tools/updateDerivativesTest.php:
```diff
@@ -0,0 +1,360 @@
+        $db =& \Database::singleton();
+
+        //Add derivative folder to filepath
+        $physioFilePath = "bids_imports/__BIDSVersion_1.2.2/sub-DCC0001/"
```
Hard coded

AlexandraLivadas replied:
@christinerogers This is a test script I'm using to test the functionalities I added and will be removed before merging.

christinerogers reviewed (changes requested):
Really nice to see this effort, thank you @AlexandraLivadas 
A few early comments on the script - requesting changes, see comments below. 
Also to do for this PR: 
- [ ] Add to the module readme: Downloads packages for this module (and for DQT e.g.) need regular verification/updating via a script called ___. To set up, configure, and run it (including via cronjob), here are the steps: ...   (roughly 2-4 sentences)

ridz1208 commented on SQL/New_patches/2021-04-26_add_annotation_data.sql:
```diff
@@ -0,0 +1,143 @@
+--SQL patch to create and populate physiological annotation tables
+
+DROP TABLE IF EXISTS `physiological_annotation_instance`;
+DROP TABLE IF EXISTS `physiological_annotation_parameter`;
+DROP TABLE IF EXISTS `physiological_annotation_archive`;
+DROP TABLE IF EXISTS `physiological_annotation_file`;
+DROP TABLE IF EXISTS `physiological_annotation_file_type`;
+DROP TABLE IF EXISTS `physiological_annotation_label`;
```
these drops need to be added to the 9999 schema

AlexandraLivadas replied:
These were added in PR #7345. As mentioned in this PR description, this is a patch that I am using to add data to the database for testing purposes and it will be deleted before merging this PR.

ridz1208 commented on modules/electrophysiology_browser/php/sessions.class.inc:
```diff
@@ -45,29 +48,43 @@ class Sessions extends \NDB_Page
+     *
+     * @return bool whether the user can create and edit the derivatives
+     */
+    function _hasEditDerivPerms(\User $user): bool
```
same

ridz1208 commented on modules/electrophysiology_browser/php/sessions.class.inc:
```diff
@@ -448,6 +481,591 @@ class Sessions extends \NDB_Page
+     * @return void
+     * @throws SodiumException
+     */
+    static function updateDerivativeFiles(int $physioFileID): void
```
I'm not too involved in the EEG design usually so this comment might be inaccurate.

This function does a ton of SQL selects, some even in a loop of instances, which is a good way to waste a lot of time. Is there no way to compile everything that needs to be queried and then run a fraction of the queries to get all the needed data ?

ridz1208 commented on modules/electrophysiology_browser/php/sessions.class.inc:
```diff
@@ -448,6 +481,591 @@ class Sessions extends \NDB_Page
+            $params
+        );
+
+        if ($this->_hasEditDerivPerms($user)) {
```
you are building and running queries above this point which will have no effect (as far as i can see) if the user doesn't have the permission, that's inefficient.

You should consider checking for permission as early as possible and returning without wasting time

ridz1208 commented on modules/electrophysiology_browser/php/sessions.class.inc:
```diff
@@ -448,6 +481,591 @@ class Sessions extends \NDB_Page
+            'AnnotationInstanceID' => $values['instance_id']
+        ];
+
+        if ($this->_hasEditDerivPerms($user)) {
```
same comment as above, you don't need do do a lot of this if your user doesn't have the proper permission. do your permission check first

ridz1208 reviewed (changes requested):
Requested changes on the SQL side to fit LORIS standards

On the code side, I only did  review for red flags as I would not be able to comment on annotations and EEG logic.

AlexandraLivadas commented on modules/electrophysiology_browser/php/sessions.class.inc:
```diff
@@ -448,6 +481,591 @@ class Sessions extends \NDB_Page
+     *
+     * @return array array with the data for each annotation for the file
+     */
+    function getAnnotations(int $physioFileID): array
```
@laemtl This method will be part of the initial GET request to view the annotations in the frontend (similar to `getDownloadLinks` above) so I followed the same formatting and made this method public. Let me know if this makes sense!

laemtl commented on modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc:
```diff
@@ -0,0 +1,535 @@
+            //Get data from POST request
+            $metadata = [
+                'Description' => $metadata['description'],
+                'Sources'     => $metadata['sources'],
```
Shouldn't this be $electrophysioFile?

AlexandraLivadas replied:
No. In the BIDS specifications there are two fields: `Sources` and `IntendedFor`. IntendedFor is defined as: `REQUIRED. Link to the data file for which the annotations are intended.` and  Sources is defined as `RECOMMENDED. Relative link or description of the file(s) used to make the annotations.`. For the IntendedFor entry, I link this data to a physiological file ID in the database.

laemtl replied:
Following our discussion, I think 'Sources' is not necessary and 'this can be removed.

laemtl commented on modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc:
```diff
@@ -0,0 +1,535 @@
+            $metadata = [
+                'Description' => $metadata['description'],
+                'Sources'     => $metadata['sources'],
+                'Author'      => $metadata['author']
```
Shouldn't this be $user?

laemtl commented on modules/electrophysiology_browser/php/models/electrophysioannotations.class.inc:
```diff
@@ -0,0 +1,535 @@
+
+            //Get data from POST request
+            $metadata = [
+                'Description' => $metadata['description'],
```
We already have a description field below. Should this one be generic?

christinerogers commented on modules/electrophysiology_browser/jsx/components/electrophysiology_session_panels.js:
```diff
@@ -548,19 +564,43 @@ class FilePanel extends Component {
+   */
+  showLoadingMessage() {
+      let type = 'success';
+      let title = 'We are currently working hard to download your files';
+      let text = 'Please be patient 😴';
```
```suggestion
      let title = 'Your download will begin in a few moments';
      let text = 'Thank you for your patience';
```

christinerogers replied:
Hi @AlexandraLivadas  - Could you accept this change to the messaging ? thanks, 
Christine

christinerogers reviewed (changes requested):
tweaks below.  
thanks for your efforts here.

cmadjar commented:
@AlexandraLivadas to work on this.

christinerogers commented:
Hi @AlexandraLivadas - could you remind me, what's the status here?  just waiting for free time to work on this or a new maintainer - is that about the size of it?

AlexandraLivadas commented:
> Hi @AlexandraLivadas - could you remind me, what's the status here? just waiting for free time to work on this or a new maintainer - is that about the size of it?

Hi @christinerogers. I will review right now but I think the only thing left here is to configure the cron job and then update the documentation with instructions.

christinerogers reviewed (changes requested):
Hi Alex , 
I just added some comments on your latest commit (instead of on the whole PR) -
Let me know if you have any questions --

AlexandraLivadas commented:
> "Updating File Download packages" i think is more accurate to say, perhaps ?

@christinerogers I don't think this would make it clear, since **all** derivative files are updated with the script, whereas only the files that are being downloaded are updated when the download button is pressed.

christinerogers commented:
> > "Updating File Download packages" i think is more accurate to say, perhaps ?
> 
> @christinerogers I don't think this would make it clear, since **all** derivative files are updated with the script, whereas only the files that are being downloaded are updated when the download button is pressed.

Ok, spell these 2 different update processes out so it's clear for the module documentation.

christinerogers commented:
> cut this info about what a cronjob is. assume the user knows.

@AlexandraLivadas I don't think this comment has been addressed.  Please cut all the info introducing what a cronjob is, what the settings mean, etc.

christinerogers commented:
> > > "Updating File Download packages" i think is more accurate to say, perhaps ?
> > 
> > 
> > @christinerogers I don't think this would make it clear, since **all** derivative files are updated with the script, whereas only the files that are being downloaded are updated when the download button is pressed.
> 
> Ok, add just enough clarification to let an admin know how the file updates (can) happen in each case - so it's clear for the module documentation.

Has this comment been addressed?  I think you re-requested my review after my comment.

AlexandraLivadas commented:
@christinerogers This is ready for re-review :)

christinerogers commented on modules/electrophysiology_browser/README.md:
````diff
@@ -41,23 +41,9 @@ electrode information, task event information, the actual recording) -- as well
-```
-
-For example, this commmand will run the script on the first day of every month at midnight.   
+When annotations are added or edited in the browser, updates are maded in the database but not directly in the associated derivative files for this data. These derivative files must also be updated. When a download is slected in the browser, the derivative files for the selected dataset are updated before downloading is triggered. All derivative files in the file system can be updated using the [ ...
````
proofreading / please correct: 
maded 
directly -> immediately
slected -> selected
downloading is triggered -> the actual download process starts
to simplify, Replace "All derivative files in the file system can be updated using the..." with "A script is provided to perform this [script name and link."

christinerogers commented on modules/electrophysiology_browser/README.md:
```diff
@@ -41,23 +41,9 @@ electrode information, task event information, the actual recording) -- as well
-For example, this commmand will run the script on the first day of every month at midnight.   
+When annotations are added or edited in the browser, updates are maded in the database but not directly in the associated derivative files for this data. These derivative files must also be updated. When a download is slected in the browser, the derivative files for the selected dataset are updated before downloading is triggered. All derivative files in the file system can be updated using the [ ...
+
+This commmand will run the script on the first day of every month at midnight.   
```
mmm

christinerogers replied:
```suggestion
Below is a sample crontab entry to execute this script *every night* at midnight  
```
Please also change the cronjob fields so that it's not first of the month, but every night

christinerogers reviewed (changes requested):
Minor updates and typo fixes, please and thank you :) 
slack me when it's ready for re-review 
C

christinerogers reviewed (changes requested):
One change request you missed: 
> downloading is triggered -> the actual download process starts

and -- my apologies, please scratch the word "immediately" (no adverb needed there i don't think)

AlexandraLivadas commented:
@ridz1208 @laemtl This is ready for re-review!

christinerogers commented:
@AlexandraLivadas this PR is critical this week.  Could you rebase and let me know what next steps are? 

cc: @laemtl

christinerogers reviewed (approved):
Approving.  
@AlexandraLivadas are you still working on TestPlan updates for this PR per @cmadjar 's last review?

cmadjar reviewed (changes requested):
@AlexandraLivadas the test plan still needs to be edited to add manual testing step for the new permission and the download capabilities of the annotation.

AlexandraLivadas commented:
@cmadjar I have added instructions to test the download button for annotations. The permission I have added will only be relevant to test once there is frontend functionality for editing annotations. Let me know if the changes I made look good!

cmadjar reviewed (approved):
Thank you @AlexandraLivadas! LGTM.

driusan commented on SQL/New_patches/2021-02-19_electrophysiology_annotation_tables.sql:
```diff
@@ -1,5 +1,10 @@
-
+DROP TABLE IF EXISTS `physiological_annotation_instance`;
+DROP TABLE IF EXISTS `physiological_annotation_parameter`;
+DROP TABLE IF EXISTS `physiological_annotation_archive`;
```
These drop tables should not be in patches for existing projects. It adds an unnecessary risk of data loss.
